### PR TITLE
Bugfix/5950

### PR DESCRIPTION
**Pull request by richp405**
##### SUMMARY
bugfix for jira 5950
The changes behavior of job_info to return a 'notfound' record, instead of a runtime exception, when a requested job is not found.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
sub-features of job.py

##### ADDITIONAL INFORMATION
get_job_output  and get_job_info were both firing a runtime exception when a specified job was not found, which would interrupt a playbook, potentially unnecessarily.

### CODE REVIEW

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:95]: When the job string is passed in for a job that does not exist , I wonder if any of these `NOTFOUND` could be populated. Its likely you won't have that info but I am unsure without trying it out. 
Ie subsystem & system

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:96]: Using I don't mind the use of `NOTFOUND` but we need to normalize/standardize on either use of spaces, case and verbosity. 

Let me further explain, some of the other properties have more verbose messages like `Job not found` where others are just `NOTFOUND`, `NOTFND`, so this is the verbosity difference I see and think case and use of spaces is pretty self explanatory.

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:113]: Seeing inconsistencies where NOTFND vs NOTFOUND used.

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:100]: I question what we should have here for a job that does not exist in terms of ret_code, 

By the definition of the property:

- msg - would would be the job error such as: "JCL ERROR" | "ABEND(S|U)xxxx" | "CONV ABEND" | "SEC ERROR" | "SYS FAIL"| CANCELED"| "CC xxxx"

- msg_code - would the string value of the job error reduced down to something more consumable such as "00" ("00" | "04" | "08" | "S0C4"
- msg_txt  - is a place holder for us such that if we can deduce from the msg_code some message that is of value and known to us but does not come from z/OS or ZOAU we can inject it here to help users. For example:
"The target is unreachable"
"Error code xxxx received, please check yyyy is installed and configured"
- code - if and only if we can convert msg_code into an integer value we return it otherwise we return None
Something like msg_code :"00" can be converted to type int: code : 00
Something like msg_code :"ABEND", msg_code :"ABEND(S1234)", msg_code :"CANCELED", can not convert thus return None

JSON mockup for ret_code:
          "ret_code":{    
"msg":" CANCELED",  // "JCL ERROR" | "ABEND(S|U)xxxx" | "CONV ABEND" | "SEC ERROR" | "SYS FAIL"| CANCELED"| "CC xxxx"     
"msg_code":"00",        // "00" | "04" | "08" | "S0C4"      
"msg_txt":"Some message we can deduce from the return code",      
"code":04   
          }
          
 We can discuss, my thoughts are to keep it aligned to the above definitions so maybe:
 msg = 'JOB NOT FOUND'
 msg_txt = 'Unable to find a job matching {job_name} and {job_id}'
 code = None
 msg_code = "None" //unsure of this idea but can't think of a better suggestion.

File: plugins/modules/zos_job_output.py
Comment (ddimatos) [inline plugins/modules/zos_job_output.py:83]: Maybe something like this:
description:
     List of the jobs output or a single NOTFOUND record when the job has not been run. 
     
     
`NOTFOUND` could change depending on my earlier comments

File: plugins/modules/zos_job_query.py
Comment (ddimatos) [inline plugins/modules/zos_job_query.py:80]: Similar to other comment maybe slightly more verbose

The list of the z/OS job(s) and status  or a single NOTFOUND record when the job has not been run.

File: plugins/modules/zos_job_submit.py
Comment (ddimatos) [inline plugins/modules/zos_job_submit.py:127]: Maybe something like this:
description:
List of the jobs output or a single NOTFOUND record when the job has not been run.

NOTFOUND could change depending on my earlier comments

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py]: argh!  Yep...before I corrected the doc... the code was there.  I've committed the correction.

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:92]: This code is triggered when there is no return for the given JobID.

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:85]: I like the thought.  I've added it there, but tweaked the response to indicate 'no jobs found' (meaning ANY) as opposed to a specific job not being found.

File: plugins/modules/zos_job_output.py
Comment (richp405) [reply plugins/modules/zos_job_output.py:83]: I can see that

Comment (richp405) [reply plugins/modules/zos_job_output.py:83]: I changed the entries to:
    Returns:
        list[dict] -- The output information for a list of jobs matching specified criteria.
        If no job status is found, this will return an empty job code with msg=Job not found

Let me know if that needs more tweaking

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:110]: https://ibmmainframes.com/references/a21.html  steep name can be 8 characters, max, just like jobname.  So, it could be NOTFOUND.  I'll change to that, so we can see if that helps.

File: plugins/modules/zos_job_output.py
Comment (ddimatos) [reply plugins/modules/zos_job_output.py:83]: Looks good , thanks @richp405

File: plugins/modules/zos_job_query.py
Comment (ddimatos) [reply plugins/modules/zos_job_query.py:80]: @richp405 - seems that this one has been missed yet the other one is fixed
https://github.com/ansible-collections/ibm_zos_core/pull/260#discussion_r629592125

File: plugins/modules/zos_job_submit.py
Comment (ddimatos) [reply plugins/modules/zos_job_submit.py:127]: @richp405 - seems that this one has been missed yet the other one is fixed or you are waiting for approval before updating others similar comments
https://github.com/ansible-collections/ibm_zos_core/pull/260#discussion_r629592125

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py]: so, in job not found, should owner be "NONE" or None?

File: plugins/modules/zos_job_submit.py
Comment (richp405) [reply plugins/modules/zos_job_submit.py:127]: I forgot about submit.

File: plugins/module_utils/job.py
Comment (ddimatos) [reply plugins/module_utils/job.py]: This works for me, as we discussed this probably is better than 'none' to reduce some of the ambiguity of none.

Comment (richp405) [reply plugins/module_utils/job.py]: system, subsystem, owner are now all none.  Step name within not found job is now blank.

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:97]: Lets upper case this if you agree to match how we usually see an an error from z/OS for consistency.

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:39]: I know I am being overly critical here but "Job not found" is not equal to "Job Not Found", one has uppercase Not and Found.

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:110]: If the Job is not found, would it be enough to just go with `None` for stepname? What is your thought @richp405

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:133]: See earlier comment about the msg "Job not found" not being equal to "Job Not Found"

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:116]: The message case should be consistent with others, see prior message.  I am not referring to the fact that this is a plural use error, but the uppercase Not and Found in other messages. Error messages should all be standardized for users to parse on.

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:115]: This error case (the plural one) seems to be missing the "job["ret_code"]["code"] = None" property, maybe I am misunderstanding why it was left off or maybe it is a coding error.

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:359]: See other messages about message case sensitivity above for `msg=Job not found`

File: plugins/modules/zos_job_output.py
Comment (ddimatos) [inline plugins/modules/zos_job_output.py:83]: I don't believe this description is accurately telling our users the outcome.

If a job is not found (assuming by name or id - I have to trace to code to see if its both) that the result would return a return code 00 and a msg "Job Not Found" or "No jobs found" , we don't return a NOTFOUND anymore for a job but it seems you still have it code for a job stepname which I commented below maybe we can just use "None" for a job step name when the job is not found. Either way I think we have some work to do here in the description/explanation. 

Couple of other things, maybe was you write the update you will see maybe why I was hesitant on supporting the plural an singular use case because now we need to instruct users with two different outcomes of a message depending on if its plural or not (depends which module is executing the code, i understand the plural use case) but lets talk about this. 

To be honest I am not even sure we should return a 00, but maybe because the module did not fail we should) for job["ret_code"]["msg_code"] = "00"

File: plugins/modules/zos_job_query.py
Comment (ddimatos) [inline plugins/modules/zos_job_query.py:75]: I wonder if this doc is actually just copy/paste because in a query module (I need to check the code) do we ever change anything to cause the state to change?

File: plugins/modules/zos_job_query.py
Comment (ddimatos) [inline plugins/modules/zos_job_query.py:81]: See other comments about case sensitivity for msg=Job not found.

File: plugins/modules/zos_job_submit.py
Comment (ddimatos) [inline plugins/modules/zos_job_submit.py:128]: Case sensitivity mentioned elsewhere for msg=Job not found.

File: plugins/modules/zos_job_submit.py
Comment (ddimatos) [inline plugins/modules/zos_job_submit.py:127]: See my other comment , it appears this doc also needs to be updated, similar to my other comment.

File: plugins/modules/zos_job_submit.py
Comment (ddimatos) [inline plugins/modules/zos_job_submit.py:616]: Can we change the case to more sentence structure (upper and lower case) since we have other edits now?

File: tests/functional/modules/test_zos_job_output_func.py
Comment (ddimatos) [inline tests/functional/modules/test_zos_job_output_func.py:100]: Case sensitivity here as well if we change the source.

File: tests/functional/modules/test_zos_job_output_func.py
Comment (ddimatos) [inline tests/functional/modules/test_zos_job_output_func.py:90]: How are you handling the plural version of your error message for query , "No jobs found"

File: plugins/module_utils/job.py
Comment (ddimatos) [inline plugins/module_utils/job.py:99]: Lets chat about this, because according to our contract/doc , we should return None for msg_code , see below "if and only if we can convert msg_code into an integer value we return it otherwise we return None" 

- msg_code - would the string value of the job error reduced down to something more consumable such as "00" ("00" | "04" | "08" | "S0C4"

- code - if and only if we can convert msg_code into an integer value we return it otherwise we return None
Something like msg_code :"00" can be converted to type int: code : 00
Something like msg_code :"ABEND", msg_code :"ABEND(S1234)", msg_code :"CANCELED", can not convert thus return None

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:97]: agreed

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:39]: I get it ... consistency.

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:110]: If we're sticking with pythonic solutions (NoneType), I agree.

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:133]: yep

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:116]: k

File: plugins/modules/zos_job_submit.py
Comment (richp405) [reply plugins/modules/zos_job_submit.py:127]: I've tweaked it

File: plugins/modules/zos_job_submit.py
Comment (richp405) [reply plugins/modules/zos_job_submit.py:616]: done

File: tests/functional/modules/test_zos_job_output_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_output_func.py:100]: Thanks...that's the stuff that trips me up.

File: tests/functional/modules/test_zos_job_output_func.py
Comment (richp405) [reply tests/functional/modules/test_zos_job_output_func.py:90]: I'm not sure how to replicate that... it should only trigger if NO jobs are returned on the global query... while it's a possibility, I haven't seen it happen... any time I've run it there has been SOMETHING in the JES Queue.

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:99]: I think we agree code should be None, and the question is should msg_code be a simple integer?  Would changing the msg_code to "NOT FOUND" make sense?  I don't recall a length limit on that field, and that would become a string, not an integer, similar to 'SOC4".

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:115]: the plural case comes after the singular case, which already set the ret_code/code to None

**Review by ddimatos — APPROVED** [review]
Looks great @richp405 - took some work but we got it where we wanted, thanks for your patience.

**Pull request merged**
